src/coreclr/jit/importer.cpp

```diff
@@ -4134,31 +4136,36 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 // to `true` or `false`
                 // e.g., `typeof(int).IsValueType` => `true`
                 // e.g., `typeof(Span<int>).IsByRefLike` => `true`
-                if (impStackTop().val->IsCall())
+                CORINFO_CLASS_HANDLE hClass = NO_CLASS_HANDLE;
+                if (gtIsTypeof(impStackTop().val, &hClass))
                 {
-                    GenTreeCall* call = impStackTop().val->AsCall();
-                    if (call->gtCallMethHnd == eeFindHelper(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE))
+                    switch (ni)
                     {
-                        assert(call->gtArgs.CountArgs() == 1);
-                        CORINFO_CLASS_HANDLE hClass =
-                            gtGetHelperArgClassHandle(call->gtArgs.GetArgByIndex(0)->GetEarlyNode());
-                        if (hClass != NO_CLASS_HANDLE)
+                        case NI_System_RuntimeType_get_IsActualEnum:
+                        case NI_System_Type_get_IsEnum:
                         {
-                            switch (ni)
-                            {
-                                case NI_System_Type_get_IsValueType:
-                                    retNode = gtNewIconNode(eeIsValueClass(hClass) ? 1 : 0);
-                                    break;
-                                case NI_System_Type_get_IsByRefLike:
-                                    retNode = gtNewIconNode(
-                                        (info.compCompHnd->getClassAttribs(hClass) & CORINFO_FLG_BYREF_LIKE) ? 1 : 0);
-                                    break;
-                                default:
-                                    NO_WAY("Intrinsic not supported in this path.");
-                            }
-                            impPopStack(); // drop CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE call
+                            CorInfoType infoType = info.compCompHnd->getTypeForPrimitiveValueClass(hClass);
+                            retNode              = gtNewIconNode(
+                                (eeIsValueClass(hClass) &&
+                                 // getTypeForPrimitiveNumericClass seems to not normalize enums
+                                 info.compCompHnd->getTypeForPrimitiveNumericClass(hClass) == CORINFO_TYPE_UNDEF &&
+                                 // we need to check for void here
+                                 infoType != CORINFO_TYPE_UNDEF && infoType != CORINFO_TYPE_VOID)
```

Comment:
Would it be better to just have a method on JIT/EE interface for this?

Also, this won't do the right thing for RuntimeTypes where IsEnum and IsActualEnum return different results. The JIT probably won't ever see RuntimeTypes like that here, but it can change in future and it is hard to audit.

Reply:
Agreed. We should do this analysis over on the runtime side.

Reply:
> Would it be better to just have a method on JIT/EE interface for this?
> 
> Also, this won't do the right thing for RuntimeTypes where IsEnum and IsActualEnum return different results. The JIT probably won't ever see RuntimeTypes like that here currently, but it can change in future and it is hard to audit.

@MichalPetryka waiting for you to follow up on moving this logic into a new interface method.

Reply:
> > Would it be better to just have a method on JIT/EE interface for this?
> > Also, this won't do the right thing for RuntimeTypes where IsEnum and IsActualEnum return different results. The JIT probably won't ever see RuntimeTypes like that here currently, but it can change in future and it is hard to audit.
> 
> @MichalPetryka waiting for you to follow up on moving this logic into a new interface method.

Sorry, I had COVID again and then was quite busy for the past few weeks, will get to it this weekend.

Reply:
> > @MichalPetryka waiting for you to follow up on moving this logic into a new interface method.
> 
> Sorry, I had COVID again and then was quite busy for the past few weeks, will get to it this weekend.

No apologies needed. Just wanted to (gently) remind you about this ...